### ISSUE TITLE
Batch Size Finder: Allow datamodule to decide fate when OOM thrown with new hook

### ISSUE THREAD

**Issue by SeanNaren** [Lightning-AI/lightning#13402]
## 🚀 Feature

Related #10243 (PR did not get to the finish line).

### Motivation

After speaking to "Harry" on the Lightning Slack channel, we came up with some improvements that can be made to the batch size finder. This spawned from a discussion around automatic accumulation in real-time, but ended up being closer to the developer experience when using the batch size finder before training.

### Pitch

We should allow the DataModule to control what happens when the batch size finder resets the dataloaders. This is required when certain data-loaders need to be reset in a particular manner (order by size of examples, such that the largest examples are first).

```python
class MyDataModule(pl.LightningDataModule)
    ...
    def on_batch_size_finder_reset(self, model, trainer):
        trainer.reset_train_dataloader(model)
        trainer.reset_val_dataloader(model)

```

This function would be called via the batch size finder when a reset needs to be made, i.e:
https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/tuner/batch_size_scaling.py#L149-L150

or 

https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/tuner/batch_size_scaling.py#L188-L189

We should also ensure that we reset when an OOM is raised, as this is omitted from the exception clause: https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/tuner/batch_size_scaling.py#L195

cc @rohitgr7 @carmocca

**Comment by mogwai** [comment]
I think this would also require some documentation on when and why you might want to override this method in the docs with a link in sections mentioning batch_finding flags?

**Comment by justusschock** [comment]
If this becomes a hook in the DM, it should probably also be part of the LM, as you can also user LM hooks to define the loaders

**Comment by carmocca** [comment]
This would be fixed by having it as a callback that the user can easily modify cc @rohitgr7

**Comment by rohitgr7** [comment]
not really, but we can expose it if required. That would require few follow-ups after that PR get's merged since currently the callback is using existing code to better review.

A current workaround can be:
```py
def train_dataloader(self):
    if self.trainer.state.fn == TrainerFn.TUNING:
        # do stuff
    ...
```
